tpm2tools/keys_test.go

```diff
@@ -70,3 +72,184 @@ func BenchmarkStorageRootKeyRSA(b *testing.B) {
+	if err != nil {
+		t.Fatalf("failed to unseal: %v", err)
+	}
+	if !bytes.Equal(secret, unseal) {
+		t.Fatalf("unsealed (%v) not equal to secret (%v)", unseal, secret)
+	}
+
+	err = tpm2.PCRExtend(rwc, tpmutil.Handle(7), tpm2.AlgSHA256, extension, "")
+	if err != nil {
+		t.Fatalf("failed to extend pcr: %v", err)
+	}
+
+	// unseal should not succeed.
+	unseal, err = key.Unseal(rwc, []int{7}, priv, pub)
+	if err == nil {
+		t.Fatalf("unseal should have caused an error: %v", err)
+	}
+}
+
+func TestComputeSessionAuth(t *testing.T) {
+	rwc := internal.GetTPM(t)
+	defer rwc.Close()
+
+	pcrVal, err := tpm2.ReadPCR(rwc, 7, tpm2.AlgSHA256)
+	if err != nil {
+		t.Fatalf("failed to read pcr: %v", err)
+	}
+	pcrs := map[int][]byte{7: pcrVal}
+
+	pcrNums := make([]int, 0, len(pcrs))
+	for key := range pcrs {
+		pcrNums = append(pcrNums, key)
+	}
+
+	getAuth, err := getPCRSessionAuth(rwc, pcrNums)
+	if err != nil {
+		t.Fatalf("failed to get session auth: %v", err)
+	}
+
+	computeAuth, err := computePCRSessionAuth(rwc, pcrs)
+	if err != nil {
+		t.Fatalf("failed to compute session auth: %v", err)
+	}
+
+	if !bytes.Equal(computeAuth, getAuth) {
+		t.Fatalf("computed auth (%v) not equal to session auth(%v)", computeAuth, getAuth)
+	}
+}
+
+func TestSelfReseal(t *testing.T) {
+	rwc := internal.GetTPM(t)
+	defer rwc.Close()
+
+	key, err := StorageRootKeyRSA(rwc)
+	if err != nil {
+		t.Fatalf("can't create srk from template: %v", err)
+	}
+
+	secret := []byte("test")
+
+	if err != nil {
+		t.Fatalf("failed to extend pcr: %v", err)
+	}
+
+	priv, pub, err := key.Seal(rwc, []int{7}, secret)
+	if err != nil {
+		t.Fatalf("failed to seal: %v", err)
+	}
+
+	unseal, err := key.Unseal(rwc, []int{7}, priv, pub)
+	if err != nil {
+		t.Fatalf("failed to unseal: %v", err)
+	}
+	if !bytes.Equal(secret, unseal) {
+		t.Fatalf("unsealed (%v) not equal to secret (%v)", unseal, secret)
+	}
+
+	pcrVal, err := tpm2.ReadPCR(rwc, 7, tpm2.AlgSHA256)
+	if err != nil {
+		t.Fatalf("failed to read pcr: %v", err)
+	}
+
+	pcrs := map[int][]byte{7: pcrVal}
+
+	priv, pub, err = key.Reseal(rwc, pcrs, priv, pub)
+	if err != nil {
+		t.Fatalf("failed to reseal: %v", err)
+	}
+
+	unseal, err = key.Unseal(rwc, []int{7}, priv, pub)
+	if err != nil {
+		t.Fatalf("unseal failed: %v", err)
+	}
+}
+func TestReseal(t *testing.T) {
+	rwc := internal.GetTPM(t)
+	defer rwc.Close()
+
+	key, err := StorageRootKeyRSA(rwc)
+	if err != nil {
+		t.Fatalf("can't create srk from template: %v", err)
+	}
+
+	secret := []byte("test")
+
+	if err != nil {
+		t.Fatalf("failed to extend pcr: %v", err)
+	}
+
+	priv, pub, err := key.Seal(rwc, []int{7}, secret)
+	if err != nil {
+		t.Fatalf("failed to seal: %v", err)
+	}
+
+	unseal, err := key.Unseal(rwc, []int{7}, priv, pub)
+	if err != nil {
+		t.Fatalf("failed to unseal: %v", err)
+	}
+	if !bytes.Equal(secret, unseal) {
+		t.Fatalf("unsealed (%v) not equal to secret (%v)", unseal, secret)
+	}
+
+	// extend pcr8 as a hack to get the intended value of pcr7 post extension.
```

Comment:
This commend is confusing, why are two PCRs ever used if we're only ever sealing/unsealing on a single PCR?

It might be more useful to just do the extension purely in software. Perhaps as an operation on some PCRValue struct.

Reply:
I could write the rest of predictive sealing, but I considered it out of scope for this change.

Reply:
Split into a helper for clarity.